### PR TITLE
K8SPXC-428 Integration with vault-token-issuer

### PR DESCRIPTION
**Pull request by blez**
[K8SPXC-428](https://jira.percona.com/browse/K8SPXC-428)

This PR is making PXC operator work with vault-token-issuer.
The main idea is to stop running if "issue-vault-token" annotation exists and start only after it disappeared. If annotation disappeared that's mean that new secret with vault config was issued. 
Basically that's all. After annotation disappeared cluster will be restarted (PXC pods will be scaled down to 0 size and scaled up back to initial cluster size). Restart will be done automatically because of hash of vault secret will be changed in appropriate annotation.
Restart is needed to make PXC use new config and understand that encryption was turned on.

### CODE REVIEW

File: pkg/controller/pxc/upgrade.go [hunk 1]
```diff
@@ -54,22 +54,30 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
...
 	}
 	if sslInternalHash != "" && cr.CompareVersionWith("1.1.0") >= 0 {
 		currentSet.Spec.Template.Annotations["percona.com/ssl-internal-hash"] = sslInternalHash
 	}
 
+	vaultConfigHash, err := r.getSecretHash(cr, cr.Spec.VaultSecretName, true)
+	if err != nil {
+		return fmt.Errorf("upgradePod/updateApp error: update secret error: %v", err)
+	}
+	if vaultConfigHash != "" && cr.CompareVersionWith("1.1.0") >= 0 {
```
Comment (delgod) [inline pkg/controller/pxc/upgrade.go:77]:
```suggestion
	if vaultConfigHash != "" && cr.CompareVersionWith("1.6.0") >= 0 {
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| scaling | passed |
| init-deploy | passed |
| big-data | passed |
| storage | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| limits | passed |
| upgrade | failed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| storage | passed |
| init-deploy | passed |
| big-data | passed |
| scaling | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| limits | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Review by delgod — APPROVED** [review]

File: deploy/cr.yaml [hunk 2]
```diff
@@ -6,6 +6,8 @@ metadata:
     - delete-pxc-pods-in-order
 #    - delete-proxysql-pvc
 #    - delete-pxc-pvc
+#  annotations:
+#    issue-vault-token: "yes"
```
[the remark this answers is not in the document]
Comment (blez) [reply deploy/cr.yaml]: https://github.com/percona/percona-xtradb-cluster-operator/pull/543/commits/931567cc53753a15d22e19e44ced52f343e5736e

File: e2e-tests/build-and-run [hunk 3]
```diff
@@ -7,8 +7,8 @@ test_dir="$(dirname $0)"
 src_dir="$(realpath $test_dir/..)"
 
 if [[ $BUILD != "0" ]]; then
-    echo "building, set BUILD=0 to suppress this"
-    $test_dir/build
```
Comment (qJkee) [inline e2e-tests/build-and-run:11]: afaik we are using spaces everywhere, pls revert

File: pkg/controller/pxc/controller.go [hunk 4]
```diff
@@ -170,6 +170,13 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
+
+	// wait untill token issued to run PXC in data encrypted mode.
+	if ok := o.ShouldWaitForTokenIssue(); ok {
```
Comment (qJkee) [inline pkg/controller/pxc/controller.go:175]:
```
if o.ShouldWaitForTokenIssue() {
	log.Info("wait for token issuing")
	return rr, nil
}
```

File: pkg/controller/pxc/controller.go [hunk 5]
```diff
@@ -488,22 +495,30 @@ func (r *ReconcilePerconaXtraDBCluster) deploy(cr *api.PerconaXtraDBCluster) err
...
 	}
 	if sslInternalHash != "" && cr.CompareVersionWith("1.1.0") >= 0 {
 		nodeSet.Spec.Template.Annotations["percona.com/ssl-internal-hash"] = sslInternalHash
 	}
 
+	vaultConfigHash, err := r.getSecretHash(cr, cr.Spec.VaultSecretName, true)
+	if err != nil {
+		return fmt.Errorf("upgradePod/updateApp error: update secret error: %v", err)
+	}
+	if vaultConfigHash != "" && cr.CompareVersionWith("1.1.0") >= 0 {
```
Comment (qJkee) [inline pkg/controller/pxc/controller.go:518]: Should version be `1.6.0`?

Comment (blez) [reply pkg/controller/pxc/controller.go:518]: Already fixed

File: e2e-tests/build-and-run (hunk 3 above)
Comment (blez) [reply e2e-tests/build-and-run]: done

File: pkg/controller/pxc/controller.go (hunk 4 above)
Comment (blez) [reply pkg/controller/pxc/controller.go]: done

File: pkg/controller/pxc/controller.go (hunk 5 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go:518]: but, it's still here =)

Comment (delgod) [reply pkg/controller/pxc/controller.go:518]:
```suggestion
	if vaultConfigHash != "" && cr.CompareVersionWith("1.6.0") >= 0 {
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| init-deploy | passed |
| recreate | passed |
| big-data | passed |
| storage | passed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing | passed |
| monitoring | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | failed |
| upgrade-consistency | passed |
| affinity | passed |
| security-context | passed |
| one-pod | passed |
| demand-backup | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | failed |
| users | passed |
| operator-self-healing | passed |
| demand-backup-encrypted-with-tls | failed |

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| big-data | passed |
| init-deploy | passed |
| scaling | passed |
| storage | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| limits | passed |
| monitoring | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| security-context | passed |
| affinity | passed |
| one-pod | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| big-data | passed |
| scaling | passed |
| scaling-proxysql | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| security-context | passed |
| demand-backup-encrypted-with-tls | failed |

**Pull request merged**